plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -332,8 +332,16 @@ else if (identity.getId() == TRINO_MERGE_PARTITION_DATA) {
                     }
                 });
 
+        TupleDomain<IcebergColumnHandle> dynamicFilterTupleDomain = dynamicFilter.getCurrentPredicate().transformKeys(IcebergColumnHandle.class::cast);
+        if (dynamicFilterTupleDomain.getDomains().isPresent() && !(dynamicFilterTupleDomain.isAll() || dynamicFilterTupleDomain.isNone())) {
+            // Filter out the partition filters from the dynamic filter because they are irrelevant at this stage
+            dynamicFilterTupleDomain = TupleDomain.withColumnDomains(dynamicFilterTupleDomain.getDomains().get()
+                    .entrySet().stream()
+                    .filter(e -> !partitionKeys.containsKey(e.getKey().getId()))
```

Comment:
why not use `io.trino.plugin.iceberg.IcebergUtil#canEnforceConstraintWithinPartitioningSpec`?

Reply:
The filter `!partitionKeys.containsKey(e.getKey().getId())` is concise enough and easy to reason about.